File: c++/0438-find-all-anagrams-in-a-string/solution1.cpp
#include <cstring>
#include <string>
#include <vector>

using namespace std;

class Solution1 {
    bool areEquals(int* a, int* b) {
        for (int i = 0; i < 26; ++i) {
            if (a[i] != b[i]) {
                return false;
            }
        }

        return true;
    }

public:
    vector<int> findAnagrams(string s, string p) {
        int sSize = s.size();
        int pSize = p.size();
        if (sSize < pSize) {
            return {};
        }

        int pCounts[26] = {0};
        for (int i = 0; i < pSize; ++i) {
            ++pCounts[p[i] - 'a'];
        }

        vector<int> result;

        int sCounts[26] = {};
        for (int i = 0; i <= sSize - pSize; ++i) {
            for (int j = i; j < i + pSize; ++j) {
                ++sCounts[s[j] - 'a'];
            }

            if (areEquals(sCounts, pCounts)) {
                result.push_back(i);
            }

            memset(sCounts, 0, 26 * sizeof(int));
        }

        return result;
    }
};

File: java/0438-find-all-anagrams-in-a-string/Solution1.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution1 {
    public List<Integer> findAnagrams(String s, String p) {
        if (s.length() < p.length()) {
            return new ArrayList<>();
        }

        int[] pCounts = new int[26];
        for (char c : p.toCharArray()) {
            ++pCounts[c - 'a'];
        }

        List<Integer> result = new ArrayList<>();

        int[] sCounts = new int[26];
        for (int i = 0; i <= s.length() - p.length(); ++i) {
            for (int j = i; j < i + p.length(); ++j) {
                ++sCounts[s.charAt(j) - 'a'];
            }

            if (Arrays.equals(sCounts, pCounts)) {
                result.add(i);
            }

            Arrays.fill(sCounts, 0);
        }

        return result;
    }
}
